### DISCUSSION TITLE
[@types/inquirer] prompt type is not correct for Reactive interface

### DISCUSSION THREAD

**Discussion by ryanjwessel** [DefinitelyTyped/DefinitelyTyped#55467] [Issues with a @types package]
[According to the `inquirer` docs](https://github.com/SBoudrias/Inquirer.js#reactive), one can use `Rx.Subject` to dynamically add questions to a prompt. 

However, when I follow the instructions in that `README.md`, I get the following TypeScript error:

```ts
var prompts = new Rx.Subject();
inquirer.prompt(prompts); // Leads to below error
Argument of type 'Subject<unknown>' is not assignable to parameter of type 'QuestionCollection<unknown>'.
  Type 'Subject<unknown>' is missing the following properties from type 'Observable<DistinctQuestion<unknown>>': _isScalar, _trySubscribe, _subscribets(2345)
```

Here are my current versions: 
```
    "@types/inquirer": "^7.3.3",
    "inquirer": "^8.1.2",
    "rxjs": "^7.3.0",
```

Is Inquirer v8 not compatible with my installed version of Inquirer's types?

[a comment by an automated account is not shown]

**Comment by neurolag** [comment]
Hey, @ryanjwessel
Though, the test-files in this repository don't include tests on using `inquirer` with react (as far as I can see), the current type-declarations support `rxjs`.

However, you need to change your code a tiny bit: You need to pass a generic parameter to the `Rx.Subject` in order to let the typescript-compiler know that your `rxjs`-subject is actually working with inquirer-questions.

Long story short - your code has to either look like this:
```ts
import inquirer = require("inquirer");
import Rx = require("rxjs");

let prompts = new Rx.Subject<inquirer.DistinctQuestion>(); // Notice the `inquirer.DistinctQuestion`-portion.
inquirer.prompt(prompts);
```

or like this:
```ts
import inquirer = require("inquirer");
import Rx = require("rxjs");

let prompts: Rx.Subject<inquirer.DistinctQuestion> = new Rx.Subject(); // Once more, notice the `inquirer.DistinctQuestion`-portion.
```

Hope this will help you solving your issue.

**Reply by ryanjwessel** [reply]
Hi @manuth, thanks for your reply.

When I tried this change, I encountered a different TypeScript error:

```ts
let prompts: Rx.Subject<inquirer.DistinctQuestion> = new Rx.Subject(); // Once more, notice the `inquirer.DistinctQuestion`-portion.
inquirer.prompt(prompts);

Argument of type 'Subject<DistinctQuestion<Answers>>' is not assignable to parameter of type 'QuestionCollection<unknown>'.
  Type 'Subject<DistinctQuestion<Answers>>' is missing the following properties from type 'Observable<DistinctQuestion<unknown>>': _isScalar, _trySubscribe, _subscribets(2345)
```

**Reply by neurolag** [reply]
Do you have `rxjs` `v6.4.0` or higher installed?
The `rxjs` in the `inquirer`-types was updated a few days ago

**Reply by ryanjwessel** [reply]
Yeah, I have `rxjs@7.3.0` installed.

**Reply by ryanjwessel** [reply]
I tried providing a type parameter of `Answer` to the `prompt` function to resolve the type error:

```ts
let prompts = new Subject<DistinctQuestion>();
prompt<Answers>(prompts);
// which leads to the below error
Argument of type 'Subject<DistinctQuestion<Answers>>' is not assignable to parameter of type 'QuestionCollection<Answers>'.
  Type 'Subject<DistinctQuestion<Answers>>' is missing the following properties from type 'Observable<DistinctQuestion<Answers>>': _isScalar, _trySubscribe, _subscribets(2345)
```

Then, when I dug into the types for `prompt`, it seems like `QuestionCollection` is missing `Subject<DistinctQuestion<T>>` as a type? I copied these from my `node_modules/@types/inquirer/index.d.ts`.

```ts
       /**
         * Prompts the questions to the user.
         */
        <T>(questions: QuestionCollection<T>, initialAnswers?: Partial<T>): Promise<T> & { ui: PromptUI };

    /**
     * Represents a collection of questions.
     *
     * @template T
     * The type of the answers.
     */
    type QuestionCollection<T extends Answers = Answers> =
        | DistinctQuestion<T>
        | ReadonlyArray<DistinctQuestion<T>>
        | Observable<DistinctQuestion<T>>;
```

**Reply by neurolag** [reply]
Having `Subject<DistinctQuestion<T>>` wouldn't solve this issue, really, as `Subject<DistinctQuestion<T>>` extends `Observable<DistinctQuestion<T>>` - which means it's already included.

As far as I can see, `rxjs@6` has some crappy type-definitions (it exposes internal variables, namely `_isScalar`, `_trySubscribe` and `_subscribe`).
`rxjs@7` seems to have fixed this issue and thus `rxjs@6`s `Observable` is incompatible with `rxjs@7`s `Observable`.

Honestly I'm not quite sure what to do at this point - we could either go and downgrade `@types/inquirer`s `rxjs` again (back to whatever it was before) or upgrade it to `rxjs@7` and have fingers crossed, that `rxjs@6`s `Observable` is assignable to `rxjs@7`s `Observable`

**Reply by ryanjwessel** [reply]
Thanks for your help on this @manuth. I see now that my version of `rxjs`, `7.3.0`, was mismatched with `@types/inquirer`'s version of `rxjs`, which was `6.4.0`.  

When I downgraded my project's version of `rxjs` to match, this problem was fixed. 

It seems like it would be good to upgrade to the latest version of `rxjs` for this project, but I am not sure of the level-of-effort required for this, as I just started working with it today. 😄

**Reply by neurolag** [reply]
I just had a look at the current version of `rxjs`
Looks like the current version, 7.3.0, has (different) internal variables in their type-declarations as well.
Looks like we can't upgrade `@types/inquirer`s `rxjs`-dependency until `inquirer` pushes an update.
